support/releaseinfo/registryclient/client.go

```diff
@@ -223,36 +232,110 @@ func ExtractImageFilesToDir(ctx context.Context, imageRef string, pullSecret []b
+		return nil, reference.DockerImageReference{}, fmt.Errorf("failed to parse image reference %q: %w", imageRef, err)
 	}
 	repo, err := registryContext.Repository(ctx, ref.DockerClientDefaults().RegistryURL(), ref.RepositoryName(), false)
 	if err != nil {
-		return nil, nil, fmt.Errorf("failed to create repository client for %s: %w", ref.DockerClientDefaults().RegistryURL(), err)
+		return nil, reference.DockerImageReference{}, fmt.Errorf("failed to create repository client for %s: %w", ref.DockerClientDefaults().RegistryURL(), err)
 	}
-	firstManifest, location, err := manifest.FirstManifest(ctx, ref, repo)
+	return repo, ref, nil
+}
+
+// Gets the manifest from a docker image
+func GetManifest(ctx context.Context, imageRef string, pullSecret []byte) (distribution.Manifest, error) {
+	repo, ref, err := GetRepoSetup(ctx, imageRef, pullSecret)
+
+	var srcDigest digest.Digest
+	if len(ref.ID) > 0 {
+		srcDigest = digest.Digest(ref.ID)
+	} else if len(ref.Tag) > 0 {
+		desc, err := repo.Tags(ctx).Get(ctx, ref.Tag)
+		if err != nil {
+			return nil, err
+		}
+		srcDigest = desc.Digest
+	} else {
+		return nil, err
+	}
+	manifests, err := repo.Manifests(ctx)
 	if err != nil {
-		return nil, nil, fmt.Errorf("failed to obtain root manifest for %s: %w", imageRef, err)
+		return nil, err
 	}
-	_, layers, err := manifest.ManifestToImageConfig(ctx, firstManifest, repo.Blobs(ctx), location)
+	digestsManifest, err := manifests.Get(ctx, srcDigest, manifest.PreferManifestList)
 	if err != nil {
-		return nil, nil, fmt.Errorf("failed to obtain image layers for %s: %w", imageRef, err)
+		return nil, err
+	}
+
+	return digestsManifest, nil
+}
+
+// Determines whether an image is a manifest listed image and contains manifests the following processor architectures: amd64, arm64, s390x, ppc64le
+func IsMultiArchManifestList(ctx context.Context, imageRef string, pullSecret []byte) (bool, error) {
+	srcManifest, err := GetManifest(ctx, imageRef, pullSecret)
+	if err != nil {
+		return false, fmt.Errorf("failed to retrieve manifest %s: %w", imageRef, err)
+	}
+
+	mediaType, payload, err := srcManifest.Payload()
+	if err != nil {
+		return false, fmt.Errorf("failed to get payload %s: %w", imageRef, err)
+	}
+
+	// This is the media type for manifest listed (aka fat manifest) images per the docker documentation - https://docs.docker.com/registry/spec/manifest-v2-2/
+	if mediaType == "application/vnd.docker.distribution.manifest.list.v2+json" {
+		m := new(manifestlist.DeserializedManifestList)
+		if err = m.UnmarshalJSON(payload); err != nil { 
+			return false, fmt.Errorf("failed to get unmarshalled manifest list: %w", err) 
+		}
+
+		count := 0
+		for _, arch := range m.ManifestList.Manifests {
+			switch arch.Platform.Architecture {
+			case ArchitectureAMD64, ArchitectureS390X, ArchitecturePPC64LE, ArchitectureARM64:
+				count = count + 1
+			}
+		}
+
+		if count > 1 {
+			return true, nil
+		} else {
+			return false, nil
+		}
+	} else {
+		return false, nil
```

Comment:
instead of having the main logic in an if lets return early false, e.g.

```
if if mediaType =! expected {
   return false
}

Main logic goes here
```

Reply:
Updated